### PR TITLE
Add support for k8s operators behind a "k8s-operators" feature flag.

### PR DESCRIPTION
**Pull request by wallyworld**
## Checklist

 - [x ] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?

Controller only facade changes.

## Description of change

An operator charm is one with deployment mode = "operator".
It does not call k8s-spec-set - instead the charm pod is the workload.

The main changes are to the caas unit provisioner worker and k8s broker.
The worker does not start a deployment worker anymore as that deployment worker is only used to scale and handle k8s spec changes - both of these are disabled for operators. It also tweaks how it watches the cluster to report juju application and unit info - it watches the charm operator. The worker uses a new DeployentMode api call to see what type of charm it is.

The relevant k8s broker methods take a mode argument (operator or workload) and tweak the name of the resources they look for and/or the label selector used.

## QA steps

bootstrap a k8s controller, enable the feature flag
`$ juju controller-config features=[k8s-operators]`

deploy a charm with metadata:

```yaml
deployment:
  mode: operator
```
and which does not call k8s-spec-set.

juju status will show the app and a single unit as active
juju show-status-log will show all the start up hooks run
juju scale-application will result in an error
kubectl get all will show only the operator pod etc in the cluster

Check that mariadb-k8s still deploys and scales as before.

### CODE REVIEW

**Comment by wallyworld** [comment]
[POC charm as operator](https://trello.com/c/q8WH74wW/1081-poc-charm-as-operator)

File: apiserver/facades/controller/caasunitprovisioner/provisioner.go [hunk 1]
```diff
@@ -226,6 +227,45 @@ func (f *Facade) applicationScale(tagString string) (int, error) {
...
+	ch, _, err := app.Charm()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	var mode charm.DeploymentMode
+	if d := ch.Meta().Deployment; d != nil {
+		mode = d.DeploymentMode
+	}
+	if mode == "" {
+		mode = charm.ModeWorkload
```
Comment (hpidcock) [inline apiserver/facades/controller/caasunitprovisioner/provisioner.go]: This probably should have been handled by the charm lib

Comment (wallyworld) [reply apiserver/facades/controller/caasunitprovisioner/provisioner.go]: Just being defensive. charm lib does declare that workload is default

**Comment by wallyworld** [comment]
$$merge$$

File: worker/caasunitprovisioner/application_worker.go [hunk 2]
```diff
@@ -116,18 +126,20 @@ func (aw *applicationWorker) loop() error {
...
-					logger.Warningf("k8s cloud hosting %q has disappeared", aw.application)
+					logger.Warningf("k8s cloud hosting %q has disappeared", aw.application, aw.mode)
 					return nil
 				}
 				return errors.Annotatef(err, "failed to start unit watcher for %q", aw.application)
 			}
+			brokerUnitsChannel = brokerUnitsWatcher.Changes()
 		}
-		if appOperatorWatcher == nil {
+		if appOperatorWatcher == nil && aw.mode == caas.ModeWorkload {
```
Comment (ycliuhw) [inline worker/caasunitprovisioner/application_worker.go]: do we default aw.mode to `caas.ModeWorkload` if charm doesn't have it?

Comment (ycliuhw) [reply worker/caasunitprovisioner/application_worker.go]: ah just saw above conversation

**Comment by wallyworld** [comment]
$$merge$$

File: apiserver/facades/controller/caasunitprovisioner/provisioner.go (hunk 1 above)
Comment (hpidcock) [reply apiserver/facades/controller/caasunitprovisioner/provisioner.go]: IMO defensive would have been to error.

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**